packages-exp/analytics-exp/src/helpers.ts

```diff
@@ -0,0 +1,308 @@
+ */
+
+import {
+  DynamicConfig,
+  DataLayer,
+  Gtag,
+  CustomParams,
+  ControlParams,
+  EventParams,
+  MinimalDynamicConfig
+} from '@firebase/analytics-types';
+import { GtagCommand, GTAG_URL } from './constants';
+import { logger } from './logger';
+
+/**
+ * Inserts gtag script tag into the page to asynchronously download gtag.
+ * @param dataLayerName Name of datalayer (most often the default, "_dataLayer").
+ */
+export function insertScriptTag(dataLayerName: string): void {
+  const script = document.createElement('script');
+  // We are not providing an analyticsId in the URL because it would trigger a `page_view`
+  // without fid. We will initialize ga-id using gtag (config) command together with fid.
+  script.src = `${GTAG_URL}?l=${dataLayerName}`;
+  script.async = true;
+  document.head.appendChild(script);
+}
+
+/**
+ * Get reference to, or create, global datalayer.
+ * @param dataLayerName Name of datalayer (most often the default, "_dataLayer").
+ */
+export function getOrCreateDataLayer(dataLayerName: string): DataLayer {
+  // Check for existing dataLayer and create if needed.
+  let dataLayer: DataLayer = [];
+  if (Array.isArray(window[dataLayerName])) {
+    dataLayer = window[dataLayerName] as DataLayer;
+  } else {
+    window[dataLayerName] = dataLayer;
+  }
+  return dataLayer;
+}
+
+/**
+ * Wrapped gtag logic when gtag is called with 'config' command.
+ *
+ * @param gtagCore Basic gtag function that just appends to dataLayer.
+ * @param initializationPromisesMap Map of appIds to their initialization promises.
+ * @param dynamicConfigPromisesList Array of dynamic config fetch promises.
+ * @param measurementIdToAppId Map of GA measurementIDs to corresponding Firebase appId.
+ * @param measurementId GA Measurement ID to set config for.
+ * @param gtagParams Gtag config params to set.
+ */
+async function gtagOnConfig(
+  gtagCore: Gtag,
+  initializationPromisesMap: { [appId: string]: Promise<string> },
+  dynamicConfigPromisesList: Array<
+    Promise<DynamicConfig | MinimalDynamicConfig>
+  >,
+  measurementIdToAppId: { [measurementId: string]: string },
+  measurementId: string,
+  gtagParams?: ControlParams & EventParams & CustomParams
+): Promise<void> {
+  // If config is already fetched, we know the appId and can use it to look up what FID promise we
+  /// are waiting for, and wait only on that one.
+  const correspondingAppId = measurementIdToAppId[measurementId as string];
+  try {
+    if (correspondingAppId) {
+      await initializationPromisesMap[correspondingAppId];
+    } else {
+      // If config is not fetched yet, wait for all configs (we don't know which one we need) and
+      // find the appId (if any) corresponding to this measurementId. If there is one, wait on
+      // that appId's initialization promise. If there is none, promise resolves and gtag
+      // call goes through.
+      const dynamicConfigResults = await Promise.all(dynamicConfigPromisesList);
```

Comment:
We should use `Promise.allSettled()` instead of `Promise.all()`, because we should proceed to try to find a match even if some promises rejected.

`Promise.allSettled()` is a pretty new addition to the es spec, so we need to use a polyfill. We can create one ourselves:

```
const a = Promise.resolve(1);
const b = Promise.reject(new Error(2));
const c = Promise.resolve(3);

// allSettled polyfill code
Promise.all([a, b, c].map(p => p.catch(e => e)))
  .then(results => console.log(results)) // 1,Error: 2,3
```

Reply:
I made a polyfill function. It adds an extra tick to the async stack so this same function has to be used in tests (see helpers.test.ts).